### ISSUE TITLE
upgrading to terraform 0.12 requires _all_ dependent modules to be upgraded

### ISSUE THREAD

**Issue by deitch** [hashicorp/terraform#21698]
Are all terraform modules in git and the terraform registry that have not been upgraded to 0.12 unusable? I had thought it was at least somewhat backwards compatible.

### Terraform Version

```
Terraform v0.12.1
```

### Terraform Configuration Files

Any tf config that references a module which is not yet upgraded to 0.12. For example, will use [official Hashicorp aws consul module from terraform registry](https://registry.terraform.io/modules/hashicorp/consul/aws/0.6.1) :

```hcl
variable "vpc" {}
variable "subnets" {
  type = list(string)
}
terraform {
  required_version = ">= 0.12"
}

provider "aws" {
}

module "consul" {
  source  = "hashicorp/consul/aws//modules/consul-cluster"
  version = "0.6.1"

  allowed_inbound_cidr_blocks = ["10.0.0.0/8"]

  ami_id = "ami-0087f029c88fa4619" # Amazon Linux 2

  # Add this tag to each node in the cluster
  cluster_tag_key   = "consul-cluster"
  cluster_tag_value = "mycluster"

  cluster_name  = "mycluster"
  instance_type = "t3.small"
  vpc_id        = var.vpc
  cluster_size  = 3
  subnet_ids    = var.subnets
  ssh_key_name  = "default"
  user_data = "#cloud-config"  # this can be filled in in any way, doesn't matter as we never get there
}
```

### Expected Behavior

tf 0.12 would tolerate 0.11 HCL, at least for submodules. 

### Steps to Reproduce

Use the sample tf above and run `tf init` and `tf plan`.

### Additional Context

Terraform 0.12 is stricter about language requirements. Definitely understood. And it is ok to enforce it on my direct code.

However, if my code includes other modules - and it should, since recommended best practice is not to "recreate the world" - then I cannot start using tf 0.12 until every single module on which I am dependent, and every one of theirs, all the way down the chain, also upgrades. This invalidates the vast majority of modules in the terraform registry.

If there are enough steps down, it is highly likely that some module somewhere down the chain will not be upgraded for a very long time.

Am I prevented from switching to 0.12 until _every_ module is upgrade?

**Comment by apparentlymart** [comment]
Hi @deitch,

The answer to this question depends on the details of the module. For modules that use only documented features from Terraform 0.11 and prior, many modules can work without modifications.

Unfortunately lots of modules relied on missing validation in the 0.11 configuration parser either to write things in a non-idiomatic way or to cause Terraform to behave in a way it wasn't intended to behave as a workaround for some limitation. Regularizing the language to add the 0.12 features required fixing various parser bugs that these workarounds relied on, and so modules using those cannot be used with Terraform 0.12 until they are upgraded and the workarounds replaced with use of the new intentional features in Terraform 0.12.

Unfortunately if the modules you have used tend to be ones containing these unusual patterns then indeed they must be upgraded before you can upgrade, assuming that you wish to continue using them.

Terraform 0.11's language runtime and Terraform 0.12's language runtime are incompatible with one another, so we cannot use them together in the same configuration. We actually attempted that some later Terraform 0.11 releases and it was our original plan to support both runtimes at once, but the language is such a cross-cutting concern in the Terraform codebase that it was infeasible to do so. That left the alternatives of either not improving the language at all or providing an automatic upgrade tool to help with upgrading, and we decided to go with the second option so that we could still fix the many limitations of the 0.11 language.

A similar tradeoff applies to waiting for modules to be upgraded: we could've made the final 0.12.0 release later to wait for certain registry modules to be upgraded first, but instead we decided it was better to release it sooner (especially considering that it was already late) so that a subset of users could begin using it immediately and so that the final version would be available to give module developers confidence in testing their upgraded versions.

I'm sorry that not everything was ready to support Terraform 0.12.0 in all configurations on release day. I hope the above gives some context as to why we decided this was the least bad of the available options. Terraform Registry modules are maintained separately from Terraform Core, so if there are specific modules you are interested in which haven't yet been upgraded I'd suggest opening issues in those codebases to let the maintainers know you are interested, which may help them prioritize which modules to upgrade first.

**Comment by deitch** [comment]
Thanks for the detailed explanation @apparentlymart . 

Practically, that can create complex situations. Here are some examples.

* I have a tf configuration that depends on external modules. I want to upgrade my config to take advantage of 0.12 features (I really do, actually). However, the external modules on which I depend are relying on those "implicit API" features from pre-0.12. I am blocked from upgrading.
* I have a tf config that depends on two external modules, A and B. A upgrades to 0.12, B does not. I now have to remain on an older release of A, until B upgrades. It gets worse. A discovers a major bug, and fixes it. I cannot take advantage of the fixed bug because it is post-0.12.
* I maintain a module. I either can upgrade to 0.12 to support customers who are 0.12-compliant, or stay 0.11 to support customers who are 0.11-compliant. It is possible I will not be able to do both without maintaining 2 separate and divergent code bases.

Of course, the module problem means some modules will not be upgraded for a very long time, which means some configs will not be, forcing some modules not to be... it is an (un)virtuous loop.

Is there any path for reducing this conflict? Is there a validator that I can run that says, "this code works on both 0.11 and 0.12"?

For example, if I run `tf 0.12upgrade` on codebase that contains the following:

```terraform
property = "${var.foo}"
```

it will replace it with

```terraform
property = var.foo
```

I like the newer language much better, it is cleaner in all respects. But I cannot support both at the same time (unless you tell me that 0.11.<something> does support `var.foo` and not just `"${var.foo}"`).

Also, is there a list of "language conflicts" between the two?

**Comment by apparentlymart** [comment]
Hi @deitch!

Indeed, upgrading lots of modules maintained by different people can be particularly complicated.

For those modules that are open source, a path forward could be to try running  `terraform 0.12upgrade` on them and submit the result as a pull request to help distribute the work for the module maintainers who look after many modules. While some modules may require some additional effort, I expect that for many the automatic upgrade tool will guide through the work and the blocker has just been that there are lots of modules to update.

Creating a new major version that supports Terraform 0.12 and leaving the previous version largely frozen is the primary pattern here, which is of course unfortunate if the older version needs maintenance. While it's not true for all modules, we've found that in practice many modules tend to quickly reach a point when they are "done" and don't require significant ongoing maintenance, because once infrastructure exists we tend not to make significant changes to it. For those modules that are seeing continued ongoing maintenance there is, unfortunately, no alternative to maintaining a 0.11-compatible maintenance branch and cutting releases for the previous minor version.

If the module you maintain is using only features that are cross-compatible then it probably already works in both or could be made to do so with only minor changes. In the latter case, I was able to do this for a few modules by just working through the few errors Terraform 0.12 generated one by one and making minor changes myself to make it work. If you're not sure what fix is appropriate, running `0.12upgrade` on a throwaway branch can give you a "cheat sheet" for ways to fix particular errors, even if you don't accept all of the changes it proposes.

Running `terraform validate` with both Terraform 0.11 and 0.12 is the best way to check for cross-compatibility, if that is  your goal. Unfortunately the 0.11 configuration loader code is no longer fully functional in 0.12 (for the reasons I was mentioning before as to why we couldn't ship them both in a single release) so we don't  have any single tool for verifying both at once.

**Comment by deitch** [comment]
Hi @apparentlymart . Yup, I get it from the technology perspective. I am just looking at it as a user, realizing that you left me only "big-bang upgrades". Personally, I have written enough modules that I am fine opening PRs (if I have time, which is a different issue entirely), but most users want things to work. 

0.12 brings so much (long-overdue) goodness, for people to be told, "you cannot use it until absolutely everything you depend upon, and all their dependencies, and....". From a product perspective, it would have been great if there were a transitional version, e.g. 0.12.x supported both, while 0.13+ support only the new language, i.e. a deprecation path.

> we've found that in practice many modules tend to quickly reach a point when they are "done" and don't require significant ongoing maintenance, because once infrastructure exists we tend not to make significant changes to it.

Indeed, that is precisely the point. Those mature modules likely have very few resources on them, so no one to grab it and say, "I will make this work under 0.12 and cut a new release", while those are the very ones that have been around so long that many configurations and other modules depend upon them.

Ah, well, I am making the argument from the product management perspective - deprecate then EOL, don't suddenly switch, support users' flows - but I am not about to change HC's views on the matter, am I? :-)
